tools/monitoring/yaml-handler/parser.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// The yamlHandler is responsible for fetching, parsing config yaml file. It also allows user to
+// retrieve a particular record from the yaml.
+
+package yamlHandler
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"regexp"
+)
+
+type AlertConditions struct {
+	JobNameRegex string `yaml:"job-name-regex"`
+	Occurrences  int
+	JobsAffected string `yaml:"jobs-affected"`
+	PrsAffected  string `yaml:"prs-affected"`
+	Period       int
+}
+
+type PatternSpec struct {
+	ErrorPattern string `yaml:"error-pattern"`
+	Hint         string
+	Alerts       []AlertConditions
+}
+
+type YamlFile struct {
+	Spec []PatternSpec `yaml:"spec"`
+}
+
+type Output struct {
+	Hint         string
+	Occurrences  int
+	JobsAffected string
+	PrsAffected  string
+	Period       int
+}
+
+//GetSpec gets the spec for a particular error pattern and a matching job name pattern
+func GetSpec(f YamlFile, pattern, jobName string) (output Output, noMatchError error) {
+	noMatchError = errors.New(fmt.Sprintf("No spec found for pattern[%s] and jobName[%s]", pattern, jobName))
+	for _, patternSpec := range f.Spec {
+		if pattern == patternSpec.ErrorPattern {
+			noMatchError = errors.New(fmt.Sprintf("Spec found for pattern[%s], but no match for job name[%s]", pattern, jobName))
```

Comment:
can we use fmt.Errorf() instead of errors.New()?

Reply:
by returning the error instead of stop the flow. We give the user of the function freedom to handle the error. It is possible that some jobs are not matched with any config intentionally.